parser/parser.y

```diff
@@ -4548,16 +4548,104 @@ SelectLockOpt:
 		$$ = ast.SelectLockInShareMode
 	}
 
-// See https://dev.mysql.com/doc/refman/5.7/en/union.html
 UnionStmt:
-	UnionClauseList "UNION" UnionOpt SelectStmt
+	UnionClauseList "UNION" UnionOpt "SELECT" SelectStmtOpts SelectStmtFieldList OrderByOptional SelectStmtLimit SelectLockOpt
 	{
+		sel := &ast.SelectStmt {
```

Comment:
Can we wrap these code into a function, and reuse it in `select` statement and here.

Reply:
Yeah, that's a better way, I'm working on this. But the modification in `selectStmt` bring in some  
 `shift/reduce`, `reduce/reduce` conflicts, I'm trying to figure them out.

Reply:
From MySQL document https://dev.mysql.com/doc/refman/5.7/en/union.html: 
> To use an ORDER BY or LIMIT clause to sort or limit the entire UNION result, parenthesize the individual SELECT statements and place the ORDER BY or LIMIT after the last one. The following example uses both clauses:
>
> (SELECT a FROM t1 WHERE a=10 AND B=1) UNION (SELECT a FROM t2 WHERE a=11 AND B=2) ORDER BY a LIMIT 10;
>
> A statement without parentheses is equivalent to one parenthesized as just shown.


Maybe we should also consider the `LIMIT` clause.

Reply:
@zz-jason 
1.  `LIMIT` was done in this pr, too. 
2. In MySQL, `LIMIT`, `ORDER BY` is not allowed in the first few `SelectStmt`. Maybe we should handle this too ? 
```sql
mysql> CREATE TABLE IF NOT EXISTS `t` (
    ->   `id` int(6) unsigned NOT NULL,
    ->   `tn` int(3) unsigned NOT NULL,
    ->   `type` varchar(200) NOT NULL,
    ->   PRIMARY KEY (`id`,`tn`)
    -> ) DEFAULT CHARSET=utf8;
Query OK, 0 rows affected (0.05 sec)

mysql> INSERT INTO `t` (`id`, `tn`, `type`) VALUES
    ->   ('1', '5', 'cat'),
    ->   ('2', '6', 'dog'),
    ->   ('3', '7', 'duck'),
    ->   ('4', '8', 'chicken');
Query OK, 4 rows affected (0.02 sec)
Records: 4  Duplicates: 0  Warnings: 0
mysql> select * from t limit 1 offset 2 union select 1,2,3 from dual;
ERROR 1221 (HY000): Incorrect usage of UNION and LIMIT
mysql> select * from t order by id  union select 1,2,3 from dual;
ERROR 1221 (HY000): Incorrect usage of UNION and ORDER BY
````

Reply:
Sure!  But may be we can dealing with issue in the next PR since this one is already large enough?